multicluster/cmd/multicluster-controller/clusterclaim_webhook.go

```diff
@@ -0,0 +1,106 @@
+
+// ClusterClaim validator
+type clusterClaimValidator struct {
+	Client    client.Client
+	decoder   *admission.Decoder
+	namespace string
+}
+
+// Handle handles admission requests.
+func (v *clusterClaimValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	clusterClaim := &mcv1alpha2.ClusterClaim{}
+	err := v.decoder.Decode(req, clusterClaim)
+	if err != nil {
+		klog.ErrorS(err, "Error while decoding ClusterClaim", "ClusterClaim", req.Namespace+"/"+req.Name)
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	switch req.Operation {
+	case admissionv1.Create:
+		if clusterClaim.Name != mcv1alpha2.WellKnownClusterClaimClusterSet && clusterClaim.Name != mcv1alpha2.WellKnownClusterClaimID {
+			return admission.Denied(fmt.Sprintf("name %s is not valid, only 'id.k8s.io' and 'clusterset.k8s.io' are valid names for ClusterClaim\n", clusterClaim.Name))
+		}
+	case admissionv1.Update:
+		oldClusterClaim := &mcv1alpha2.ClusterClaim{}
+		if req.OldObject.Raw != nil {
+			if err := json.Unmarshal(req.OldObject.Raw, &oldClusterClaim); err != nil {
+				klog.ErrorS(err, "Error while decoding old ClusterClaim", "ClusterClaim", klog.KObj(clusterClaim))
+				return admission.Errored(http.StatusBadRequest, err)
+			}
+			if oldClusterClaim.Value != clusterClaim.Value {
+				klog.ErrorS(err, "the field 'value' is immutable", "ClusterClaim", klog.KObj(clusterClaim))
+				return admission.Denied("the field 'value' is immutable")
+			}
+			return admission.Allowed("")
+		}
+	case admissionv1.Delete:
+		clusterSetList := &mcv1alpha1.ClusterSetList{}
+		if err := v.Client.List(context.TODO(), clusterSetList, client.InNamespace(v.namespace)); err != nil {
+			klog.ErrorS(err, "Error reading ClusterSet", "Namespace", v.namespace)
+			return admission.Errored(http.StatusPreconditionFailed, err)
+		}
+		deny := false
+		var existingClusterSet mcv1alpha1.ClusterSet
+		if len(clusterSetList.Items) > 0 {
+			// ClusterSet webhook will guarantee that there is only one ClusterSet at most in a given Namespace.
+			existingClusterSet = clusterSetList.Items[0]
+			klog.InfoS("value", "value", clusterClaim.Value, "clsutersetname", existingClusterSet.Name, "req.name", req.Name)
+			if clusterClaim.Value == existingClusterSet.Name || clusterClaim.Value == existingClusterSet.Spec.Leaders[0].ClusterID {
+				deny = true
+			} else {
+				for _, member := range existingClusterSet.Spec.Members {
+					if clusterClaim.Value == member.ClusterID {
+						deny = true
+						break
+					}
+				}
+			}
+		}
+		// Deny ClusterClaim deletion if the ClusterClaim is referred in a ClusterSet.
+		if deny {
+			klog.ErrorS(err, "Error deleting ClusterClaim, it's referred by a ClusterSet", "ClusterClaim", klog.KObj(clusterClaim), "ClusterSet", klog.KObj(&existingClusterSet))
```

Comment:
"The ClusterClaim is referred by a ClusterSet. Cannot delete it"

Reply:
I feel better to log only the Claim/Set names, rather than the whole object which is quite unnecessary.

Reply:
Message updated. `klog.KObj` only returns an object contains Namespace and Name, so the purpose of `klog.KObj` is to print Name and Namespace, not the whole object.